File: VB/Library/WeatherHourlyData.vb
﻿Imports System
Imports System.Collections.Generic

Namespace Library

    ''' <summary>
    ''' The WeatherHourlyData class represents hourly weather data. It encapsulates all relevant meteorological details
    ''' temperature, feelslike, humidity, dew, precip, precipprob, snow, snowdepth, preciptype, windgust, windspeed, winddir,
    ''' pressure, visibility, cloudcover, solarradiation, solarenergy, uvindex, conditions, icon, stations, and source for a specific hour.
    ''' </summary>
    Public Class WeatherHourlyData
        Private _datetime As TimeSpan?
        Private _datetimeEpoch As Long?
        Private _temp As Double?
        Private _feelslike As Double?
        Private _humidity As Double?
        Private _dew As Double?
        Private _precip As Double?
        Private _precipprob As Double?
        Private _snow As Double?
        Private _snowdepth As Double?
        Private _preciptype As List(Of String)
        Private _windgust As Double?
        Private _windspeed As Double?
        Private _winddir As Double?
        Private _pressure As Double?
        Private _visibility As Double?
        Private _cloudcover As Double?
        Private _solarradiation As Double?
        Private _solarenergy As Double?
        Private _uvindex As Double?
        Private _conditions As String
        Private _icon As String
        Private _stations As List(Of String)
        Private _source As String

        ' Constructor
        Public Sub New()
            _preciptype = New List(Of String)()
            _stations = New List(Of String)()
        End Sub

        ' Properties
        Public Property Datetime() As TimeSpan?
            Get
                Return _datetime
            End Get
            Set(ByVal value As TimeSpan?)
                _datetime = value
            End Set
        End Property

        Public Property DatetimeEpoch() As Long?
            Get
                Return _datetimeEpoch
            End Get
            Set(ByVal value As Long?)
                _datetimeEpoch = value
            End Set
        End Property

        Public Property Temp() As Double?
            Get
                Return _temp
            End Get
            Set(ByVal value As Double?)
                _temp = value
            End Set
        End Property

        Public Property Feelslike() As Double?
            Get
                Return _feelslike
            End Get
            Set(ByVal value As Double?)
                _feelslike = value
            End Set
        End Property

        Public Property Humidity() As Double?
            Get
                Return _humidity
            End Get
            Set(ByVal value As Double?)
                _humidity = value
            End Set
        End Property

        Public Property Dew() As Double?
            Get
                Return _dew
            End Get
            Set(ByVal value As Double?)
                _dew = value
            End Set
        End Property

        Public Property Precip() As Double?
            Get
                Return _precip
            End Get
            Set(ByVal value As Double?)
                _precip = value
            End Set
        End Property

        Public Property PrecipProb() As Double?
            Get
                Return _precipprob
            End Get
            Set(ByVal value As Double?)
                _precipprob = value
            End Set
        End Property

        Public Property Snow() As Double?
            Get
                Return _snow
            End Get
            Set(ByVal value As Double?)
                _snow = value
            End Set
        End Property

        Public Property SnowDepth() As Double?
            Get
                Return _snowdepth
            End Get
            Set(ByVal value As Double?)
                _snowdepth = value
            End Set
        End Property

        Public Property PrecipType() As List(Of String)
            Get
                Return _preciptype
            End Get
            Set(ByVal value As List(Of String))
                _preciptype = value
            End Set
        End Property

        Public Property WindGust() As Double?
            Get
                Return _windgust
            End Get
            Set(ByVal value As Double?)
                _windgust = value
            End Set
        End Property

        Public Property WindSpeed() As Double?
            Get
                Return _windspeed
            End Get
            Set(ByVal value As Double?)
                _windspeed = value
            End Set
        End Property

        Public Property WindDir() As Double?
            Get
                Return _winddir
            End Get
            Set(ByVal value As Double?)
                _winddir = value
            End Set
        End Property

        Public Property Pressure() As Double?
            Get
                Return _pressure
            End Get
            Set(ByVal value As Double?)
                _pressure = value
            End Set
        End Property

        Public Property Visibility() As Double?
            Get
                Return _visibility
            End Get
            Set(ByVal value As Double?)
                _visibility = value
            End Set
        End Property

        Public Property CloudCover() As Double?
            Get
                Return _cloudcover
            End Get
            Set(ByVal value As Double?)
                _cloudcover = value
            End Set
        End Property

        Public Property SolarRadiation() As Double?
            Get
                Return _solarradiation
            End Get
            Set(ByVal value As Double?)
                _solarradiation = value
            End Set
        End Property

        Public Property SolarEnergy() As Double?
            Get
                Return _solarenergy
            End Get
            Set(ByVal value As Double?)
                _solarenergy = value
            End Set
        End Property

        Public Property UvIndex() As Double?
            Get
                Return _uvindex
            End Get
            Set(ByVal value As Double?)
                _uvindex = value
            End Set
        End Property

        Public Property Conditions() As String
            Get
                Return _conditions
            End Get
            Set(ByVal value As String)
                _conditions = value
            End Set
        End Property

        Public Property Icon() As String
            Get
                Return _icon
            End Get
            Set(ByVal value As String)
                _icon = value
            End Set
        End Property

        Public Property Stations() As List(Of String)
            Get
                Return _stations
            End Get
            Set(ByVal value As List(Of String))
                _stations = value
            End Set
        End Property

        Public Property Source() As String
            Get
                Return _source
            End Get
            Set(ByVal value As String)
                _source = value
            End Set
        End Property
    End Class
End Namespace

File: Rust/Library/weather_hourly_data.rs
use std::option::Option;
use std::time::Duration;
use chrono::NaiveTime;

pub struct WeatherHourlyData {
    datetime: Option<NaiveTime>,
    datetime_epoch: Option<u64>,
    temp: Option<f64>,
    feels_like: Option<f64>,
    humidity: Option<f64>,
    dew: Option<f64>,
    precip: Option<f64>,
    precip_prob: Option<f64>,
    snow: Option<f64>,
    snow_depth: Option<f64>,
    precip_type: Option<Vec<String>>,
    wind_gust: Option<f64>,
    wind_speed: Option<f64>,
    wind_dir: Option<f64>,
    pressure: Option<f64>,
    visibility: Option<f64>,
    cloud_cover: Option<f64>,
    solar_radiation: Option<f64>,
    solar_energy: Option<f64>,
    uv_index: Option<f64>,
    conditions: Option<String>,
    icon: Option<String>,
    stations: Option<Vec<String>>,
    source: Option<String>,
}

impl WeatherHourlyData {
    pub fn new() -> WeatherHourlyData {
        WeatherHourlyData {
            datetime: None,
            datetime_epoch: None,
            temp: None,
            feels_like: None,
            humidity: None,
            dew: None,
            precip: None,
            precip_prob: None,
            snow: None,
            snow_depth: None,
            precip_type: None,
            wind_gust: None,
            wind_speed: None,
            wind_dir: None,
            pressure: None,
            visibility: None,
            cloud_cover: None,
            solar_radiation: None,
            solar_energy: None,
            uv_index: None,
            conditions: None,
            icon: None,
            stations: None,
            source: None,
        }
    }

    // Getters and setters
    pub fn datetime(&self) -> Option<NaiveTime> {
        self.datetime
    }

    pub fn set_datetime(&mut self, datetime: Option<NaiveTime>) {
        self.datetime = datetime;
    }

    pub fn datetime_epoch(&self) -> Option<u64> {
        self.datetime_epoch
    }

    pub fn set_datetime_epoch(&mut self, datetime_epoch: Option<u64>) {
        self.datetime_epoch = datetime_epoch;
    }

    pub fn temp(&self) -> Option<f64> {
        self.temp
    }

    pub fn set_temp(&mut self, temp: Option<f64>) {
        self.temp = temp;
    }

    pub fn feels_like(&self) -> Option<f64> {
        self.feels_like
    }

    pub fn set_feels_like(&mut self, feels_like: Option<f64>) {
        self.feels_like = feels_like;
    }

    pub fn humidity(&self) -> Option<f64> {
        self.humidity
    }

    pub fn set_humidity(&mut self, humidity: Option<f64>) {
        self.humidity = humidity;
    }

    pub fn dew(&self) -> Option<f64> {
        self.dew
    }

    pub fn set_dew(&mut self, dew: Option<f64>) {
        self.dew = dew;
    }

    pub fn precip(&self) -> Option<f64> {
        self.precip
    }

    pub fn set_precip(&mut self, precip: Option<f64>) {
        self.precip = precip;
    }

    pub fn precip_prob(&self) -> Option<f64> {
        self.precip_prob
    }

    pub fn set_precip_prob(&mut self, precip_prob: Option<f64>) {
        self.precip_prob = precip_prob;
    }

    pub fn snow(&self) -> Option<f64> {
        self.snow
    }

    pub fn set_snow(&mut self, snow: Option<f64>) {
        self.snow = snow;
    }

    pub fn snow_depth(&self) -> Option<f64> {
        self.snow_depth
    }

    pub fn set_snow_depth(&mut self, snow_depth: Option<f64>) {
        self.snow_depth = snow_depth;
    }

    pub fn precip_type(&self) -> &Option<Vec<String>> {
        &self.precip_type
    }

    pub fn set_precip_type(&mut self, precip_type: Option<Vec<String>>) {
        self.precip_type = precip_type;
    }

    pub fn wind_gust(&self) -> Option<f64> {
        self.wind_gust
    }

    pub fn set_wind_gust(&mut self, wind_gust: Option<f64>) {
        self.wind_gust = wind_gust;
    }

    pub fn wind_speed(&self) -> Option<f64> {
        self.wind_speed
    }

    pub fn set_wind_speed(&mut self, wind_speed: Option<f64>) {
        self.wind_speed = wind_speed;
    }

    pub fn wind_dir(&self) -> Option<f64> {
        self.wind_dir
    }

    pub fn set_wind_dir(&mut self, wind_dir: Option<f64>) {
        self.wind_dir = wind_dir;
    }

    pub fn pressure(&self) -> Option<f64> {
        self.pressure
    }

    pub fn set_pressure(&mut self, pressure: Option<f64>) {
        self.pressure = pressure;
    }

    pub fn visibility(&self) -> Option<f64> {
        self.visibility
    }

    pub fn set_visibility(&mut self, visibility: Option<f64>) {
        self.visibility = visibility;
    }

    pub fn cloud_cover(&self) -> Option<f64> {
        self.cloud_cover
    }

    pub fn set_cloud_cover(&mut self, cloud_cover: Option<f64>) {
        self.cloud_cover = cloud_cover;
    }

    pub fn solar_radiation(&self) -> Option<f64> {
        self.solar_radiation
    }

    pub fn set_solar_radiation(&mut self, solar_radiation: Option<f64>) {
        self.solar_radiation = solar_radiation;
    }

    pub fn solar_energy(&self) -> Option<f64> {
        self.solar_energy
    }

    pub fn set_solar_energy(&mut self, solar_energy: Option<f64>) {
        self.solar_energy = solar_energy;
    }

    pub fn uv_index(&self) -> Option<f64> {
        self.uv_index
    }

    pub fn set_uv_index(&mut self, uv_index: Option<f64>) {
        self.uv_index = uv_index;
    }

    pub fn conditions(&self) -> &Option<String> {
        &self.conditions
    }

    pub fn set_conditions(&mut self, conditions: Option<String>) {
        self.conditions = conditions;
    }

    pub fn icon(&self) -> &Option<String> {
        &self.icon
    }

    pub fn set_icon(&mut self, icon: Option<String>) {
        self.icon = icon;
    }

    pub fn stations(&self) -> &Option<Vec<String>> {
        &self.stations
    }

    pub fn set_stations(&mut self, stations: Option<Vec<String>>) {
        self.stations = stations;
    }

    pub fn source(&self) -> &Option<String> {
        &self.source
    }

    pub fn set_source(&mut self, source: Option<String>) {
        self.source = source;
    }
}
